core/meterer/accountledger.go

```diff
@@ -0,0 +1,364 @@
+	RecordReservationUsage(
+		ctx context.Context,
+		accountID gethcommon.Address,
+		timestampNs int64,
+		numSymbols uint64,
+		quorumNumbers []core.QuorumID,
+		params *PaymentVaultParams,
+	) error
+
+	// RecordOnDemandUsage attempts to record on-demand payment usage against the ledger's
+	// stored on-demand payment state and returns the new cumulative payment if successful.
+	//
+	// The method performs the following operations:
+	//   1. Validates that the requested quorums are enabled for on-demand usage
+	//   2. Calculates the payment amount based on symbols and pricing configuration
+	//   3. Checks if the payment would exceed the available on-demand balance
+	//   4. Updates the cumulative payment if sufficient balance exists
+	//
+	// Payment Calculation:
+	//   symbolsCharged = max(numSymbols, minNumSymbols)
+	//   paymentCharged = symbolsCharged * onDemandPricePerSymbol
+	//   newCumulative = currentCumulative + paymentCharged
+	//
+	// Parameters:
+	//   ctx: Context for cancellation and timeouts (currently unused in LocalAccountLedger)
+	//   accountID: The account identifier (used for logging and future database implementations)
+	//   numSymbols: The number of symbols to be charged
+	//   quorumNumbers: The list of quorums for this usage (must be on-demand enabled)
+	//   params: Payment vault parameters containing pricing and quorum configurations
+	//
+	// Returns:
+	//   (*big.Int, nil): The new cumulative payment amount if successful
+	//   (nil, error): If the usage cannot be accommodated:
+	//     - "quorum X not enabled for on-demand": Requested quorum doesn't support on-demand
+	//     - "insufficient ondemand payment": Would exceed available balance
+	//     - "quorum config not found for X": Missing payment configuration
+	RecordOnDemandUsage(
+		ctx context.Context,
+		accountID gethcommon.Address,
+		numSymbols uint64,
+		quorumNumbers []core.QuorumID,
+		params *PaymentVaultParams,
+	) (*big.Int, error)
+
+	// SetAccountState updates the ledger with complete account state including
+	// on-chain settings (reservations, on-demand) and local tracking data.
+	//
+	// This method completely replaces the current ledger state with the provided state.
+	// It performs deep copying of all data to ensure isolation between the provided
+	// state and the internal ledger state.
+	//
+	// Use Cases:
+	//   - Initial setup: Loading account state from on-chain data
+	//   - Synchronization: Updating local state with fresh on-chain data
+	//   - Testing: Setting up known state for test scenarios
+	//
+	// Parameters:
+	//   state: The complete account state to set. Nil fields will be initialized
+	//          with appropriate zero values (empty maps, zero big.Int, etc.)
+	//
+	// Thread Safety:
+	// This method is safe for concurrent use. It acquires an exclusive write lock
+	// for the entire operation to ensure atomic state replacement.
+	SetAccountState(state AccountState)
```

Comment:
Why would we need to set the account state on an already created account ledger? Couldn't we just discard the old account ledger and create a new one with the new data?

Reply:
good point, removed [a5581f8](https://github.com/Layr-Labs/eigenda/pull/1694/commits/a5581f871327b47cc67b1a5d670962a9dcc4fe3f)